internal/backend/gs/gs.go

```diff
@@ -301,16 +317,45 @@ func (be *Backend) openReader(ctx context.Context, h restic.Handle, length int,
 
 	objName := be.Filename(h)
 
+	if be.encryptKey != "" {
+		encryptKeyPresent = true
+	}
+	for _, key := range be.decryptKeys {
+		if key != "" {
+			decryptKeysPresent = true
+			break
+		}
+	}
+
 	be.sem.GetToken()
 
-	r, err := be.bucket.Object(objName).NewRangeReader(ctx, offset, int64(length))
+	//https://cloud.google.com/storage/docs/encryption/using-customer-supplied-keys
+	err = errors.New(("Google Storage customer-supplied encryption keys failed."))
+	if encryptKeyPresent {
+		cryptoKey, _ = base64.StdEncoding.DecodeString(be.encryptKey)
+		rangeReader, err = be.bucket.Object(objName).Key(cryptoKey).NewRangeReader(ctx, offset, int64(length))
+	}
+	if decryptKeysPresent && err != nil {
+		for _, key := range be.decryptKeys {
+
+			cryptoKey, _ = base64.StdEncoding.DecodeString(key)
+			rangeReader, err = be.bucket.Object(objName).Key(cryptoKey).NewRangeReader(ctx, offset, int64(length))
+			if err == nil {
+				break
+			}
+		}
+	}
+
+	if err != nil {
+		rangeReader, err = be.bucket.Object(objName).NewRangeReader(ctx, offset, int64(length))
```

Comment:
Please deduplicate the calls to ``be.bucket.Object(objName)` and `.NewRangeReader(ctx, offset, int64(length))`. Copy-pasting them tree times just asks for problems in the long term.

Reply:
Fixed.